### PR TITLE
Postgres

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by erikjohnston** [comment]
Things left to do:
1. Don't bother porting over the entirety of ``sent_transactions``
2. sytest reports presence failures

**Comment by erikjohnston** [comment]
> Check that the DB connections recover correctly after DB server restart (this seems to have bitten Matthew?)

We probably want to use `cp_reconnect`: http://twistedmatrix.com/documents/current/api/twisted.enterprise.adbapi.ConnectionPool.html#__init__

File: synapse/app/homeserver.py [hunk 1]
```diff
@@ -103,13 +103,11 @@ def build_resource_for_metrics(self):
             return None
 
     def build_db_pool(self):
+        name = self.db_config["name"]
+
         return adbapi.ConnectionPool(
-            "sqlite3", self.get_db_name(),
-            check_same_thread=False,
```
Comment (NegativeMjark) [inline synapse/app/homeserver.py]: The check_same_thread for sqlite3 seems to have gone missing

File: synapse/app/homeserver.py [hunk 2]
```diff
@@ -352,15 +350,46 @@ def setup(config_options):
...
+    if config.database_config:
+        with open(config.database_config, 'r') as f:
+            db_config = yaml.safe_load(f)
+    else:
+        db_config = {
+            "name": "sqlite3",
+            "args": {
+                "database": config.database_path,
+            },
+        }
```
Comment (NegativeMjark) [inline synapse/app/homeserver.py]: Would it make more sense to parse the database config in the config package?

Comment (erikjohnston) [inline synapse/app/homeserver.py]: Yes.

File: synapse/app/homeserver.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/app/homeserver.py]: Oops

**Comment by NegativeMjark** [comment]
Is it possible to migrate from postgress back to sqlite3?

File: synapse/storage/util/id_generators.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        with self._lock:
+            if self._unfinished_ids:
+                return self._unfinished_ids[0] - 1
+
+            if not self._current_max:
+                return store.runInteraction(
+                    "_compute_current_max",
+                    self._compute_current_max,
+                )
```
Comment (NegativeMjark) [inline synapse/storage/util/id_generators.py]: runInteraction will dispatch to a db thread. So self._compute_current_max will happen without the lock being held. Is this a problem?

File: synapse/app/homeserver.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/app/homeserver.py]: Fixed

File: synapse/app/homeserver.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/app/homeserver.py]: Done.

File: synapse/storage/util/id_generators.py (hunk 3 above)
Comment (erikjohnston) [inline synapse/storage/util/id_generators.py]: Fixed

**Comment by erikjohnston** [comment]
> Is it possible to migrate from postgress back to sqlite3?

There is no script to do so currently.

**Comment by NegativeMjark** [comment]
A lot of the schema files seem to use "VARCHAR(150)" for strings. Is there a good reason to pick 150? I tend towards picking 255 as an arbitrary value since that is the maximum that mysql can store using a single length byte https://dev.mysql.com/doc/refman/5.0/en/char.html.

**Comment by erikjohnston** [comment]
Using `VARCHAR(150)` was from when I was looking at MySQL support. Have changed them to use `TEXT`.

**Pull request merged**